core/src/main/scala/za/co/absa/spline/harvester/plugin/Plugin.scala

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.harvester.plugin
+
+import org.apache.spark.sql.SaveMode
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import za.co.absa.spline.harvester.builder.SourceIdentifier
+
+trait Plugin
+
+object Plugin {
+  type Params = Map[String, Any]
+  type ReadNodeInfo = (SourceIdentifier, Params)
+  type WriteNodeInfo = (SourceIdentifier, SaveMode, LogicalPlan, Params)
+
+  object Precedence {
+    final val Highest = 0
+    final val User = 1000
+    final val Normal = 5000
+    final val Lowest = Int.MaxValue
+  }
```

Comment:
AFAIK it's' a general convention. Although _JSR 250_ doesn't specify the meaning or the priority value (delegating it to the specs defining concrete applications of the annotation), but for example JAR-RS says this:
> A collection of built-in priority constants for the JAX-RS components that are supposed to be ordered based on their javax.annotation.Priority class-level annotation value when used or applied by JAX-RS runtime.
For example, JAX-RS filters and interceptors are grouped in chains for each of the message processing extension points: Pre, PreMatch, Post as well as ReadFrom and WriteTo. Each of these chains is sorted based on priorities which are represented as integer numbers. All chains, except Post, are sorted in ascending order; the lower the number the higher the priority. The Post filter chain is sorted in descending order to ensure that response filters are executed in reverse order.
JAX-RS components that belong to the same priority class (same integer value) are executed in an implementation-defined manner. By default, when the @Priority annotation is absent on a component, for which a priority should be applied, the USER priority value is used.
Since:
2.0
Author:
Marek Potociar (marek.potociar at oracle.com)

Reply:
And so does Spring (`Ordered` interface), even though in this case it isn't related to jsr 250, but still:
> The actual {@link #getOrder() order} can be interpreted as prioritization,
with the first object (with the lowest order value) having the highest
priority.